content/docs/user-guide/testing/best-practices/_index.md

```diff
@@ -0,0 +1,327 @@
+---
+linkTitle: Best Practices
+title: Test Automation Best Practices
+description: Advice for automating tests in Open 3D Engine (O3DE).
+toc: true
+weight: 600
+---
+
+This page provides advice on how many independent developers can efficiently maintain the **Open 3D Engine (O3DE)** with automated tests. The advice applies both to teams using O3DE who write their own automation, as well as to anyone contributing to O3DE. This document is not intended to serve as an exhaustive source of testing advice. It instead provides baseline definitions, and heuristics for investing in test automation. Each concept only serves as a brief introduction, as numerous in depth articles can be found published by other sources.
+
+## Avoid
+
+### I/O
+
+File access, device input, and network communication are slow operations which can create complex software and hardware dependencies. Tests performing I/O often encounter race conditions such as parallel contention. Most tests should prefer configuring mock implementations of I/O interfaces, and using in-memory buffers to simulate the behavior of actual input and output.
+
+### Sleep
+
+Do not rely on `this_thread::sleep` or `time.sleep` in tests. Be immediately suspicious if you see a sleep in a Pull Request. Some quick facts:
+
+* Tests relying on sleeps are automatically slow: yielding control to the underlying OS scheduler and returning control is a slow operation.
+  * Latency depends on the the specific device, and still can fluctuate with its physical and digital hardware state.
+* Passing a duration to a sleep function requests a *minimum* time the thread should be suspended, not an exact time it will resume.
+
+{{< caution >}}
+Avoid invoking sleep in any code, it is almost never the correct tool!
+{{< /caution >}}
+
+What are the alternatives?
```

Comment:
I think it's a good call to mention methods named "wait" which perform a similar sleep, though I also want to avoid digging too deeply into details in this document.

Reply:
we don't want to instruct users to avoid idle_wait though since its function is not exactly sleep. idle_wait_frames is sometimes required to allow the engine main thread to process updates.